### ISSUE TITLE
Unsupported PHP 8 command in Magento 

### ISSUE THREAD

**Issue by michalrostek** [magento/magento2#33680]
### Preconditions (*)

 
1. Magento 2.4.3 EE
1. PHP 7.4

### Steps to reproduce (*)
1. There is a PHP 8 (str_contains) command in \Magento\Framework\Filesystem\Directory\DenyListPathValidator:74 which is not officially supported yet ... so we have such errors during installation:

{{PHP Fatal error: Uncaught Error: Call to undefined function Magento\Framework\Filesystem\Directory\str_contains() in <...>/magento/vendor/magento/framework/Filesystem/Directory/DenyListPathValidator.php:74

This file (DenyListPathValidator) is initialized in \Magento\Framework\Filesystem\Directory\WriteFactory.php which may affect not only the above case

Expected result (*)

PHP 7 command, instead of PHP 8

Actual result (*)

PHP 8 command

**Comment by sivaschenko** [comment]
@michalrostek thanks for reporting this issue! While we are addressing the issue, there is a possible workaround: require `symfony/polyfill-php80` package via composer to the project

**Comment by sivaschenko** [comment]
The internal team is working on a fix for this, I have linked the internal Jira ticket to the GitHub issue, so the status will be updated accordingly.

**Comment by magento-engcom-team** [comment]
@sivaschenko Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by leonhelmus** [comment]
I had the same issue and because of the error i could not update my composer.lock with a composer update. Even after requiring symfony/polyfill-php80 i still get the error. During this command:

```
composer require magento/product-enterprise-edition:2.4.3 --base-magento-edition 'Commerce' --base-magento-version 2.4.2-p1 --no-update
```
and this one:
```
composer update magento/product-enterprise-edition magento/composer-dependency-version-audit-plugin sebastian/phpcpd phpstan/phpstan symfony/finder friendsofphp/php-cs-fixer allure-framework/allure-phpunit allure-framework/allure-codeception magento/magento2-functional-testing-framework --with-all-dependencies
```

This prevents us from updating the composer.lock file, because composer update rollsback after a incomplete composer update.

i get the following error:
```
Installing magento/module-inventory-catalog-search-configurable-product (1.0.0): Loading from cache
    ...Module Magento_InventoryCatalogSearchConfigurableProduct recognized as part of Magento Multi Source Inventory implementation
    ...Disabling Magento_InventoryCatalogSearchConfigurableProduct module for backward compatibility
PHP Fatal error:  Uncaught Error: Call to undefined function Magento\Framework\Filesystem\Directory\str_contains() in ~/vendor/magento/framework/Filesystem/Directory/DenyListPathValidator.php:74
```

System specifics:
M 2.4.2-p1 -> 2.4.3
php 7.4
mysql 8.0

**Comment by leonhelmus** [comment]
I took a better look and for this client we have multi source inventory disabled.

When checking the composer plugin magento/inventory-composer-installer it tries to disable all modules that have magento/module-inventory disabled, because of this it needs functionality of the magento framework which doesn't work, because str_contains does not exist for php 7.4 and this is the reason why composer.lock cannot be updated.

**Comment by mrtuvn** [comment]
~/vendor/magento/framework/Filesystem/Directory/DenyListPathValidator.php:74
I can't find this one in code base

**Comment by hostep** [comment]
@mrtuvn: the problematic code definitely exists in Magento 2.4.3, 2.4.2-p2 and 2.3.7-p1, I can see it in the pre-releases.
Since the latter two are security releases and the change discussed here is most likely fixing some security issue, those changes won't be visible here on github yet until the official release next week.

**Comment by leonhelmus** [comment]
@sivaschenko Do you already have a EAT? We would like to get started on M2.4.3 on monday if possible.

**Comment by arjunadobe** [comment]
> 
> 
> @sivaschenko Do you already have a ETA? We would like to get started on M2.4.3 on monday if possible.

As a workaround (for now, Disable all the inventory modules)

`        'Magento_Inventory' => 0,
        'Magento_InventoryAdminUi' => 0,
        'Magento_InventoryAdvancedCheckout' => 0,
        'Magento_InventoryApi' => 0,
        'Magento_InventoryBundleImportExport' => 0,
        'Magento_InventoryBundleProduct' => 0,
        'Magento_InventoryBundleProductAdminUi' => 0,
        'Magento_InventoryBundleProductIndexer' => 0,
        'Magento_InventoryCatalog' => 0,
        'Magento_InventorySales' => 0,
        'Magento_InventoryCatalogAdminUi' => 0,
        'Magento_InventoryCatalogApi' => 0,
        'Magento_InventoryCatalogFrontendUi' => 0,
        'Magento_InventoryCatalogSearch' => 0,
        'Magento_InventoryConfigurableProduct' => 0,
        'Magento_InventoryConfigurableProductAdminUi' => 0,
        'Magento_InventoryConfigurableProductFrontendUi' => 0,
        'Magento_InventoryConfigurableProductIndexer' => 0,
        'Magento_InventoryConfiguration' => 0,
        'Magento_InventoryConfigurationApi' => 0,
        'Magento_InventoryDistanceBasedSourceSelection' => 0,
        'Magento_InventoryDistanceBasedSourceSelectionAdminUi' => 0,
        'Magento_InventoryDistanceBasedSourceSelectionApi' => 0,
        'Magento_InventoryElasticsearch' => 0,
        'Magento_InventoryExportStockApi' => 0,
        'Magento_InventoryIndexer' => 0,
        'Magento_InventorySalesApi' => 0,
        'Magento_InventoryGroupedProduct' => 0,
        'Magento_InventoryGroupedProductAdminUi' => 0,
        'Magento_InventoryGroupedProductIndexer' => 0,
        'Magento_InventoryImportExport' => 0,
        'Magento_InventoryInStorePickupApi' => 0,
        'Magento_InventoryInStorePickupAdminUi' => 0,
        'Magento_InventorySourceSelectionApi' => 0,
        'Magento_InventoryInStorePickup' => 0,
        'Magento_InventoryInStorePickupGraphQl' => 0,
        'Magento_InventoryInStorePickupShippingApi' => 0,
        'Magento_InventoryInStorePickupQuoteGraphQl' => 0,
        'Magento_InventoryInStorePickupSales' => 0,
        'Magento_InventoryInStorePickupSalesApi' => 0,
        'Magento_InventoryInStorePickupQuote' => 0,
        'Magento_InventoryInStorePickupShipping' => 0,
        'Magento_InventoryInStorePickupShippingAdminUi' => 0,
        'Magento_InventoryCatalogSearchConfigurableProduct' => 0,
        'Magento_InventoryCatalogSearchBundleProduct' => 0,
        'Magento_InventoryQuoteGraphQl' => 0,
        'Magento_InventoryCache' => 0,
        'Magento_InventoryLowQuantityNotification' => 0,
        'Magento_InventoryLowQuantityNotificationApi' => 0,
        'Magento_InventoryMultiDimensionalIndexerApi' => 0,
        'Magento_InventoryProductAlert' => 0,
        'Magento_InventoryRequisitionList' => 0,
        'Magento_InventoryReservations' => 0,
        'Magento_InventoryReservationCli' => 0,
        'Magento_InventoryReservationsApi' => 0,
        'Magento_InventoryExportStock' => 0,
        'Magento_InventorySalesAdminUi' => 0,
        'Magento_InventoryGraphQl' => 0,
        'Magento_InventorySalesFrontendUi' => 0,
        'Magento_InventorySetupFixtureGenerator' => 0,
        'Magento_InventoryShipping' => 0,
        'Magento_InventoryShippingAdminUi' => 0,
        'Magento_InventorySourceDeductionApi' => 0,
        'Magento_InventorySourceSelection' => 0,
        'Magento_InventoryInStorePickupFrontend' => 0,
        'Magento_InventorySwatchesFrontendUi' => 0,
        'Magento_InventoryVisualMerchandiser' => 0,
        'Magento_InventoryWishlist' => 0,
	'Magento_InventoryLowQuantityNotificationAdminUi' => 0,
	'Magento_InventoryInStorePickupSalesAdminUi' => 0,
	'Magento_InventoryInStorePickupWebapiExtension' => 0,
	'Magento_InventoryInStorePickupMultishipping' => 0,`

**Comment by mrtuvn** [comment]
nice to know that you have name "adobe"

**Comment by leonhelmus** [comment]
It seems we now found a workaround we replaced the magento msi composer plugin and after we could update our composer.lock. So for now we can work around it 😁. So its not a huge blocker anymore! We also require polyfill80 and that also helped out! Thanks for the tip @sivaschenko !

**Comment by zhartaunik** [comment]
@leonhelmus could you clarify your idea with the replacement inventory-composer-installer? I have disabled MSI as well, but it's not clear what to do.

I did: 
```
composer require magento/inventory-composer-installer
composer require symfony/polyfill-php80
```
then ```rm -rf vendor/*```

then ```composer upgrade``` and still 
```
Fatal error: Uncaught Error: Call to undefined function Magento\Framework\Filesystem\Directory\str_contains() in /var/www/magento/vendor/magento/framework/Filesystem/Directory/DenyListPathValidator.php:74
Stack trace:
#0 /var/www/magento/vendor/magento/framework/Filesystem/Directory/CompositePathValidator.php(39): Magento\Framework\Filesystem\Directory\DenyListPathValidator->validate('/var/www/magent...', 'config.php', NULL, false)
#1 /var/www/magento/vendor/magento/framework/Filesystem/Directory/Read.php(80): Magento\Framework\Filesystem\Directory\CompositePathValidator->validate('/var/www/magent...', 'config.php', NULL, false)
#2 /var/www/magento/vendor/magento/framework/Filesystem/Directory/Write.php(359): Magento\Framework\Filesystem\Directory\Read->validatePath('config.php')
#3 /var/www/magento/vendor/magento/framework/App/DeploymentConfig/Writer.php(146): Magento\Framework\Filesystem\Directory\Write->writeFile('config.php', '<?php\nreturn [\n...')
#4 /var/www/magento/vendor/magento/inventory-composer-installer/src/D in /var/www/magento/vendor/magento/framework/Filesystem/Directory/DenyListPathValidator.php on line 74
```

The main question is why all of the Magento specialists tested this on an unsupported PHP version.

**Comment by mrtuvn** [comment]
Have you check composer.json have require package "magento/composer-dependency-version-audit-plugin". See this available at 2.4.3

**Comment by zhartaunik** [comment]
> Have you check composer.json have require package "magento/composer-dependency-version-audit-plugin". See this new available first time at 2.4.3
> Test with CE 2.4.3 php7.4. msi always come with too

Added, does not help

**Comment by leonhelmus** [comment]
Make sure that when you replace it and upate with composer that you removed your vendor directory. Otherwise the inventory composer plugin will still be called.

**Comment by zhartaunik** [comment]
@leonhelmus 
Yes, I did the following:
```
rm -rf vendor/* && composer upgrade
```
my composer.json:
```
    "require": {
...
        "composer/composer": "@alpha",
...
        "magento/composer-dependency-version-audit-plugin": "^0.1.1",
        "magento/composer-root-update-plugin": "~1.1",
        "magento/inventory-composer-installer": "^1.2",
        "magento/product-enterprise-edition": "2.4.3",
...
        "symfony/polyfill-php80": "^1.23",
...
    },
```

**Comment by leonhelmus** [comment]
Hmmm it should work. For me it does atleast.

**Comment by reense** [comment]
Adding `"symfony/polyfill-php80": "^1.23"` and disabling all MSI-related modules did the trick for me. Weird decision to use `str_contains` from PHP8 whilst not even supporting that version. 😞

**Comment by mrtuvn** [comment]
so we may possible broken func with php7.4 ? Not check yet but the code updated in 2.4-develop
https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Filesystem/Directory/DenyListPathValidator.php#L74

**Comment by mrtuvn** [comment]
@magento-engcom-team Could you assist with description of internal ticket?

**Comment by hostep** [comment]
See https://support.magento.com/hc/en-us/articles/4408021533069-Adobe-Commerce-upgrade-2-4-3-2-3-7-p1-PHP-Fatal-error-Hotfix for a hotfix patch.

**Comment by engcom-Hotel** [comment]
Adding appropriate labels in order to create a JIRA ticket.

**Comment by nathanjosiah** [comment]
This is a duplicate of https://github.com/magento/magento2/issues/33680

**Comment by nathanjosiah** [comment]
The work was already completed in https://jira.corp.magento.com/browse/MC-43059 and delivered. There are also hotfixes available for this issue.

**Comment by ihor-sviziev** [comment]
@nathanjosiah,
If it was delivered, why don't we see there changes?

**Comment by nathanjosiah** [comment]
At the moment you can view the changes in the delivered patches:

2.4.3: https://github.com/magento/magento-cloud-patches/blob/develop/patches/AC-384__Fix_Incompatible_PHP_Method__2.4.3_ce.patch
2.3.7-p1: https://github.com/magento/magento-cloud-patches/blob/develop/patches/AC-384__Fix_Incompatible_PHP_Method__2.3.7-p1_ce.patch

I see why the code isn't visible in 2.4-develop but I am trying to get an answer for why it was delivered like this.

**Comment by ihor-sviziev** [comment]
@nathanjosiah, it definitealy make sense to apply them it as a patch to already released versions, but it has to be fixed as well in the 2.4-develop branch

**Comment by nathanjosiah** [comment]
> it has to be fixed as well in the 2.4-develop branch
@ihor-sviziev It is already delivered to 2.4.4 however the way it was done the changes are not publicly available. I'm trying to get an answer for why it was done this way. 

> These patches are available only to magento cloud instances
The MQP tool pulls patches from MCP so this is already available there as well.

**Comment by mrtuvn** [comment]
https://devdocs.magento.com/guides/v2.4/release-notes/open-source-2-4-3.html
Seem the docs team already updated here

**Comment by nathanjosiah** [comment]
@ihor-sviziev 
> I'm trying to get an answer for why it was done this way.

At the time this fix was merged to 2.4.4 branches the code that it was fixing was not yet public either. It's an unfortunate combination of timing and type of change.

**Comment by ihor-sviziev** [comment]
@nathanjosiah, sorry, still didn't get. When it will be delivered to 2.4-develop branch? Should we accept https://github.com/magento/magento2/pull/33886?

**Comment by nathanjosiah** [comment]
@ihor-sviziev It won't be synced back until the 2.4.4 2022 Q1 release so I think it's fastest to process that PR. However, to prevent a merge conflict later lets make sure that PR fix matches the hotfix code.

**Comment by hostep** [comment]
The fuck? 🤦

**Comment by nathanjosiah** [comment]
> The fuck? 🤦
>
> Apparently people inside Adobe/Magento aren't familiar with cherry-picking in git?

@hostep Just plain rude.

The code that needed this fix was security-related and therefore not public until 2.4.3 was released. There is a several month gap between when 2.4.3 private security/release branches are frozen and when the 2.4.3 security code is synced back into 2.4-develop because of how our lifecycle works. 

At the time the 2.4.4 fix was delivered, the code was still private because 2.4.3 had not been delivered or synced back to 2.4-develop hence why we couldn't deliver a fix to 2.4-develop. We also couldn't deliver the fix to 2.4.3 branches because they were frozen due to our lifecycle.

Having said all that, it isn't impossible for us to port the change to 2.4-develop if we had to but it would be uncommon to go through the dev cycle twice for the same release line (2.4.4). Also, it would be much faster to process the existing PR that @ihor-sviziev mentioned, than it would be to go through the entire lifecycle internally.
